### ISSUE TITLE
Feature: Mosaic View

### ISSUE THREAD

**Issue by naglepuff** [DigitalSlideArchive/wsi-superpixel-guided-labeling#43]
There is a desire for a new view (or perhaps an enhancement to an existing view) that allows for review of many superpixels at once. 

**Potential capabilities:**
1. Ability to view several superpixels at once, potentially laid out in some sort of grid
2. Ability to group superpixels by most recent prediction/label, image, etc.
3. Ability to change the current label of a superpixel (click and drag? drop down menu? context/right-click menu)
4. Ability to sort superpixels (either the entire list or within groups)

**Questions:**
1. How should this view be accessed?
2. Who will be accessing the view?
3. What is the main use case for such a view?
4. What does a workflow that utilizes this view look like?
5. Do we care about human labels, predictions, or both?
6. What is the most important information to see in a "mosaic view?"

**Comment by jeffbaumes** [comment]
* Add chip review, by manual label as highest priority. Ideally sort/group by who made annotation, what slide, what class, where do machine and label disagree. By manual label then by slide would be a good default
* For a given chip, want to be able to show context ("minimize" the mosaic view to show the main viewer? Or perhaps we need it to be more lightweight without leaving the mosaic context).
* 9-25 at a time would be good, should be able to easily scroll or page (@cooperlab had a preference for scrolling rather than paging)
* Keep chip location stable even if relabelled, maybe show new label more boldly to indicate a change

We also discussed paper trail, be able to compare two reviewers, etc. which is out of scope for now but we should try to be forward-thinking about how we'd support that.

**Comment by cooperlab** [comment]
> **Questions:**
> 
> 1. How should this view be accessed?

In the prototype we had tabs [https://histomicsml2.readthedocs.io/en/latest/_images/main.png](https://histomicsml2.readthedocs.io/en/latest/_images/main.png)

> 2. Who will be accessing the view?

Labelers (trainees) or project supervisors.

> 3. What is the main use case for such a view?

To ensure that labeled data are correctly labeled.

> 4. What does a workflow that utilizes this view look like?

We discussed several labeling modes. In our prototype, you could move across modes and accumulate or change labeled examples that were stored in a cache. One a re-train cycle was initiated these would become permanent. We can discuss this more.

> 5. Do we care about human labels, predictions, or both?

Human labels.

> 6. What is the most important information to see in a "mosaic view?"

The instance and the class label. Here is the view that we had in the prototype [https://histomicsml2.readthedocs.io/en/latest/_images/review.png](https://histomicsml2.readthedocs.io/en/latest/_images/review.png). This prototype also shows the surrounding context, but that eats up space.

**Comment by cooperlab** [comment]
@bnmajor some initial feedback on the mosaic view from @jagstein.

Perhaps the slider can mimic the HistomicsUI w/ discrete magnification buttons.

Ideas for context include devoting part of the screen to display the context, or a quick view hover (more limited field of view).

**Comment by bnmajor** [comment]
> @bnmajor some initial feedback on the mosaic view from @jagstein.
> 
> Perhaps the slider can mimic the HistomicsUI w/ discrete magnification buttons.
> 
> Ideas for context include devoting part of the screen to display the context like in a file browser, or a quick view hover (more limited field of view).

This is great, thanks! Keeping @faiza-a in the loop as well

**Comment by bnmajor** [comment]
I've got a question about permissions: what kind of user restrictions (if any) should be considered for the mosaic view? Should only certain users be able to see the view? Or all users can view it but only certain users can "override" a label? Or no need to restrict anything? Even if this isn't something we need right away it would be good to be aware of from a design perspective.

**Comment by manthey** [comment]
> I've got a question about permissions: what kind of user restrictions (if any) should be considered for the mosaic view? Should only certain users be able to see the view? Or all users can view it but only certain users can "override" a label? Or no need to restrict anything? Even if this isn't something we need right away it would be good to be aware of from a design perspective.

I think aside from the basic permissions to view images and edit annotations we wouldn't have further permission restrictions.

However, thinking out possible other scenarios:

I think all users should be able to see the mosaic, and certainly edit labels they have set.  I could imagine that if you have a senior reviewers and a junior labelled that you might want to allow the senior to relabel/unlabel anything, but the junior not to be able to relabel something the senior labeled.  If you have multiple senior or junior editors, I'd assume that each could relabel something a peer labeled.  This seems complex to convey.

From a tracking perspective, when we are expecting multiple users, we may need to do more work.  Currently, the edit will be marked as updated by only the last editor; if there have been any edits in this epoch by a different user, maybe an edit automatically makes a new epoch.  Then, if we have the ability to see what was labelled per epoch and have forced there to be a single user per epoch, we should conceptually be able to see if a superpixel has been relabeled or unlabeled by different users.  I think we need to discuss the utility of this before we can design too much.

**Comment by cooperlab** [comment]
> I've got a question about permissions: what kind of user restrictions (if any) should be considered for the mosaic view? Should only certain users be able to see the view? Or all users can view it but only certain users can "override" a label? Or no need to restrict anything? Even if this isn't something we need right away it would be good to be aware of from a design perspective.

David has good observations below regarding training and multi-annotator scenario that we can discuss at the next meeting.

I don't find it necessary to restrict people from changing labels. The main benefit of an assigned reviewer role is to potentially track what has/not been reviewed. This type of coordination is very difficult to do off platform via emails and spreadsheets.

**Comment by bnmajor** [comment]
Added in #101

**Comment by bnmajor** [comment]
Functionality wired up in #131
